### PR TITLE
Route INFORMATION_SCHEMA queries

### LINKED ISSUE #6894: Rails: follow routing rules in information_schema.tables
_Related issue in https://github.com/vitessio/vitess/issues/6895._

#### Feature Description

Allow information_schema.tables to follow routing rules like SELECTs. This is more predictable than the current behaviour:

```
mysql> use mainunsharded;
Database changed
mysql> select * from entries limit 1;
+----+------------+----------+----------------+--------------+------------+--------+---------------------+---------------------+---------------------+------------------+-------------------------+
| id | account_id | topic_id | entryable_type | entryable_id | creator_id | status | created_at          | updated_at          | active_at           | replied_entry_id | alternative_sender_name |
+----+------------+----------+----------------+--------------+------------+--------+---------------------+---------------------+---------------------+------------------+-------------------------+
|  1 |          1 |        1 | Message        |            1 |          1 |      1 | 2019-04-08 12:49:06 | 2019-04-08 12:49:06 | 2019-04-08 12:49:06 |             NULL | NULL                    |
+----+------------+----------+----------------+--------------+------------+--------+---------------------+---------------------+---------------------+------------------+-------------------------+
1 row in set (0.11 sec)

mysql> SELECT count(*) FROM information_schema.tables WHERE table_schema = database() and table_name='entries';
+----------+
| count(*) |
+----------+
|        0 |
+----------+
1 row in set (0.10 sec)

mysql> use main;
Database changed
mysql> select * from entries limit 1;
+----+------------+----------+----------------+--------------+------------+--------+---------------------+---------------------+---------------------+------------------+-------------------------+
| id | account_id | topic_id | entryable_type | entryable_id | creator_id | status | created_at          | updated_at          | active_at           | replied_entry_id | alternative_sender_name |
+----+------------+----------+----------------+--------------+------------+--------+---------------------+---------------------+---------------------+------------------+-------------------------+
|  1 |          1 |        1 | Message        |            1 |          1 |      1 | 2019-04-08 12:49:06 | 2019-04-08 12:49:06 | 2019-04-08 12:49:06 |             NULL | NULL                    |
+----+------------+----------+----------------+--------------+------------+--------+---------------------+---------------------+---------------------+------------------+-------------------------+
1 row in set (0.10 sec)

mysql> SELECT count(*) FROM information_schema.tables WHERE table_schema = database() and table_name='entries';
+----------+
| count(*) |
+----------+
|        1 |
+----------+
1 row in set (0.10 sec)
```
and will fix an issue with Rails where it expects to be able to run [schema queries](https://github.com/rails/rails/blob/d83ee61dc6b5e73dd2853c5af9681e71c0150e30/activerecord/lib/active_record/connection_adapters/mysql/schema_statements.rb#L203-L224) to [determine data sources](https://github.com/rails/rails/blob/d83ee61dc6b5e73dd2853c5af9681e71c0150e30/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L32-L38).

#### Use Case(s)

We have both unsharded and sharded tables in different keyspaces, which we can link together with routing rules in the same database, but rails expects to be able to run schema level queries on all of them identically.

### PR DESCRIPTION
**Pull request by systay**
For queries against `INFORMATION_SCHEMA.tables`, when the user is asking for a specific table that is routed, Vitess needs to look up where the table is routed, and send the query to a shard in that keyspace.

There are two predicates we look for - `TABLE_SCHEMA = ?` and `TABLE_NAME = ?`. Vitess uses the values we are searching for to know how to route the query.

Fixes #6894

### CODE REVIEW

File: go/vt/vtgate/engine/route.go [hunk 1]
```diff
@@ -378,45 +381,99 @@ func (route *Route) paramsAllShards(vcursor VCursor, bindVars map[string]*queryp
...
+	}
+
+	return destinations, []map[string]*querypb.BindVariable{bindVars}, nil
+}
+
+func (route *Route) routeInfoSchemaQuery(vcursor VCursor, bindVars map[string]*querypb.BindVariable) ([]*srvtopo.ResolvedShard, error) {
+	defaultRoute := func() ([]*srvtopo.ResolvedShard, error) {
+		ks := route.Keyspace.Name
+		destinations, _, err := vcursor.ResolveDestinations(ks, nil, []key.Destination{key.DestinationAnyShard{}})
+		return destinations, vterrors.Wrapf(err, "failed to find information about keyspace `%s`", ks)
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/route.go:396]: should we not check for error. returning destinations and also returning an error. It this intentional?

Comment (systay) [reply go/vt/vtgate/engine/route.go:396]: yes, this is intentional. `vterrors.Wrap` will return `nil` if the error being wrapped is nil

File: go/vt/vtgate/planbuilder/system_tables.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+			if rut.eroute.SysTableTableSchema != nil {
+				return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "two predicates for table_schema not supported")
+			}
+			rut.eroute.SysTableTableSchema = out
+		} else {
+			if rut.eroute.SysTableTableName != nil {
+				return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "two predicates for table_name not supported")
+			}
+			rut.eroute.SysTableTableName = out
+		}
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/system_tables.go:44]: We do handle other expressions in the WHERE clause. When we are doing `pushFilter` on `primitiveBuilder`, we look at these expressions to do the routing, but then we still add these expressions to the query:

https://github.com/vitessio/vitess/blob/master/go/vt/vtgate/planbuilder/select.go#L310

File: go/vt/vtgate/planbuilder/testdata/filter_cases.txt [hunk 3]
```diff
@@ -1422,7 +1422,44 @@
...
+    "OperatorType": "Route",
+    "Variant": "SelectDBA",
+    "Keyspace": {
+      "Name": "main",
+      "Sharded": false
+    },
+    "FieldQuery": "select * from INFORMATION_SCHEMA.`TABLES` where 1 != 1",
+    "Query": "select * from INFORMATION_SCHEMA.`TABLES` where TABLE_SCHEMA = 'ks' or TABLE_SCHEMA = 'main'"
+  }
+}
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/filter_cases.txt:1445]: no,  for OR queries, we can't route in a good way, so we just send it to the default ks.

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**